public/services/data_source_service.ts

```diff
@@ -0,0 +1,128 @@
+  Customized,
+}
+
+const getDSMDataSourceSelectionOption = (dataSourceSelection: Map<string, DataSourceOption[]>) => {
+  const values = [...dataSourceSelection.values()];
+  // Should use default index if multi data source selected
+  if (values.length === 0 || values.length > 1 || values?.[0]?.length > 1) {
+    return null;
+  }
+  return values[0][0];
+};
+
+export class DataSourceService {
+  private dataSourceId$ = new BehaviorSubject<string | null>(null);
+  private uiSettings: IUiSettingsClient | undefined;
+  private dataSourceManagement: DataSourceManagementPluginSetup | undefined;
+  private dataSourceSelectionSubscription: Subscription | undefined;
+  private dataSourceIdFrom: DataSourceIdFrom | undefined;
+
+  constructor() {}
+
+  initDefaultDataSourceIdIfNeed() {
+    if (!this.isMDSEnabled() || this.dataSourceId$.getValue() !== null) {
+      return;
+    }
+    const defaultDataSourceId = this.uiSettings?.get('defaultDataSource', null);
+    if (!defaultDataSourceId) {
+      return;
+    }
+    this.setDataSourceId(defaultDataSourceId, DataSourceIdFrom.UiSettings);
+  }
+
+  clearDataSourceId() {
+    this.setDataSourceId(null, undefined);
+  }
+
+  getDataSourceQuery() {
+    if (!this.isMDSEnabled()) {
+      return {};
+    }
+    const dataSourceId = this.dataSourceId$.getValue();
+    if (dataSourceId === null) {
+      throw new Error('No data source id');
+    }
+    if (dataSourceId === '') {
+      return {};
+    }
+    return { dataSourceId };
+  }
+
+  isMDSEnabled() {
+    return !!this.dataSourceManagement;
+  }
+
+  setDataSourceId(newDataSourceId: string | null, dataSourceIdFrom: DataSourceIdFrom | undefined) {
+    this.dataSourceIdFrom = dataSourceIdFrom;
+    if (this.dataSourceId$.getValue() === newDataSourceId) {
+      return;
+    }
+    this.dataSourceId$.next(newDataSourceId);
+  }
+
+  getDataSourceId() {
+    return this.dataSourceId$.getValue();
+  }
+
+  subscribeDataSourceId(observer?: PartialObserver<string | null>) {
+    return this.dataSourceId$.subscribe(observer);
+  }
+
+  setup({
+    uiSettings,
+    dataSourceManagement,
+  }: {
+    uiSettings: IUiSettingsClient;
+    dataSourceManagement?: DataSourceManagementPluginSetup;
+  }) {
+    this.uiSettings = uiSettings;
+    this.dataSourceManagement = dataSourceManagement;
+
+    this.dataSourceSelectionSubscription = this.dataSourceManagement?.dataSourceSelection
+      ?.getSelection$()
+      .subscribe((dataSourceSelection) => {
+        const selectedDataSourceOption = getDSMDataSourceSelectionOption(dataSourceSelection);
+        this.setDataSourceId(
+          selectedDataSourceOption?.id ?? null,
+          DataSourceIdFrom.DataSourceSelection
+        );
+      });
+
+    this.uiSettings.get$('defaultDataSource', null).subscribe((newDataSourceId) => {
+      if (this.dataSourceIdFrom === DataSourceIdFrom.UiSettings) {
+        this.setDataSourceId(newDataSourceId, DataSourceIdFrom.UiSettings);
+      }
+    });
```

Comment:
Do we need to be aware of the `defaultDataSource`? I thought it should be transparent to plugins.

What about we initialize a hidden data source selector if none picker is mounted to the page? And that picker will help to tell us what the defaultDataSource is.

Reply:
This's a good idea. But there are three limitations after change to a transparent data source picker:
1. onSelectedDataSource in data source picker need to wait all data source fetched, it will take more time to determine the default data source.
2. The `onSelectedDataSource` won't be fired after default data source, the data source id in data source service will still use the old one.
3. Since we are depend on data source selection, the transparent data source picker will add selection value to the data source selection. We need to add some logic to filter out data source selection from this component. @raintygao may need to modify the OSD PR.
I can change to use a transparent data source picker if 2 & 3 have been addressed or skipped. But I think it would be better if we can add some defaultDataSource functions to subscribe the default data source changes in DSM plugin.

Reply:
DSM hasn't provided a way to get default data source. As it's stored in ui settings and DSM also gets this value from uisettings, so i think we don't need to depend on DSM to get this additionally. Im open to be aware of `defaultDataSource` in external plugin in this stage.

Reply:
Sure. @raintygao already raised a [PR](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6843) to add these methods in the DSM plugin. Let's keep current uiSettings implementation. Will raised another PR to fix this default data source after @raintygao 's PR merged.